[WIP] Make managedcluster reconcile and delete async

CecileRobertMichon opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind cleanup
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**: Implements #1702 for managedclusters.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2140 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Make managedcluster reconcile and delete async
```

CecileRobertMichon commented:
/cc @shysank @Jont828 @zmalik @alexeldeib @jackfrancis this should be ready for review... I realize it's a very large PR so please call me out on any scope creeps that I should try and decompose. I ended up having to a bit of refactoring of cluster and agentpool specs and moving some validation errors to webhooks so getting the spec itself wouldn't return errors. Kept the commits unsquashed for now so it should hopefully help with history and thought process.

CecileRobertMichon commented:
/retest

alexeldeib commented on exp/api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -43,10 +48,9 @@ func (m *AzureManagedControlPlane) SetupWebhookWithManager(mgr ctrl.Manager) err
-
 // Default implements webhook.Defaulter so a webhook will be registered for the type.
-func (m *AzureManagedControlPlane) Default() {
+func (m *AzureManagedControlPlane) Default(_ client.Client) {
```
where is this client coming from? isn't this the defaulter without a client?

alexeldeib commented:
overall lgtm 🥳 huge PR, nice work!

I think splitting out the custom header interface change would be nice to reduce the size of the diff? Might be possible to break up some of the scope refactoring for machinepool too?

jackfrancis reviewed (changes requested):
is the `cluster-api-provider-azure` binary meant to be committed?

CecileRobertMichon commented:
> is the cluster-api-provider-azure binary meant to be committed?

oops that was not intentional, thanks for catching that. Removed.

jackfrancis commented:
> overall lgtm 🥳 huge PR, nice work!
> 
> I think splitting out the custom header interface change would be nice to reduce the size of the diff? Might be possible to break up some of the scope refactoring for machinepool too?

+1, an extraordinary amount of work

@zmalik @michalno1 @TheNakedZealot @thiagompc are we aware of any AKS scenarios that our E2E tests don't currently cover? For reference here is the current template that we use to test:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/a8a35fc9174816a320651e2e90e135ec42d6fc66/templates/test/ci/cluster-template-prow-aks.yaml

I'm thinking that if there are any particular use-cases we aren't testing (but that folks are reliably using) we will benefit by adding those scenarios given that the next release of capz will have the changes from this PR. These changes will optimize many reconciliations, but the scope is so large we'll be more likely to avoid regressions if we have more test coverage.

jackfrancis commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,423 @@
+	// ResourceGroup is the name of the Azure resource group for this AKS Cluster.
+	ResourceGroup string
+
+	// NodeResourceGroupName is the name of the Azure resource group containing IaaS VMs.
```
I see that you changed the name of `ResourceGroupName` to `ResourceGroup` while breaking out the existing `ManagedClusterSpec` into two types. Would this be a good time to consider changing the `NodeResourceGroupName` property name to something a little more accurately descriptive? For example, `InfraResourceGroupName` (or maybe better `InfraResourceGroup` to match the omission of the `Name` suffix above.

I'm wondering if having a more general name will make things less confusing in the future when folks are using this property to do things on non-node resources in the `MC_` resource group. Also (admittedly this is exceedingly speculative) if AKS were to ever implement a "control plane-only" feature (i.e., bring your own nodes) this name would then actually be misleading.

CecileRobertMichon replied:
AKS API actually refers to it as "nodeResourceGroup" https://docs.microsoft.com/en-us/rest/api/aks/managed-clusters/get#managedcluster

properties.nodeResourceGroup | string | The name of the resource group containing agent pool nodes.
-- | -- | --

Will change it to `NodeResourceGroup` for consistency.

jackfrancis replied:
adhering to AKS's API makes the most sense 👍

CecileRobertMichon replied:
changed to `NodeResourceGroup`

sonasingh46 commented on azure/services/managedclusters/spec_test.go:
```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
nit: 2022

CecileRobertMichon commented on azure/scope/managedcontrolplane.go:
```diff
@@ -51,12 +48,15 @@ import (
 	InfraMachinePool *infrav1exp.AzureManagedMachinePool
-	MachinePool      *expv1.MachinePool
-	PatchTarget      conditions.Setter
+	OwnerMachinePool *clusterv1exp.MachinePool
```
No reason. Changed to `MachinePool`.

CecileRobertMichon commented on azure/interfaces.go:
```diff
@@ -106,6 +112,8 @@ type ResourceSpecGetter interface {
 	// ResourceGroupName returns the name of the resource group the resource is in.
 	ResourceGroupName() string
+	// CustomHeaders returns the headers that should be added to Azure API calls.
+	CustomHeaders() map[string]string
```
@alexeldeib had also suggested splitting it out. I've added a commit to add a separate interface that has `CustomHeaders()`. The downside is that it required casting in client.go before using it but it's a little cleaner since it doesn't require touching any of the other services. Let me know what you think.

sonasingh46 replied:
Yes. In separate interface -- it looks much cleaner. This should be fine to go with the PR.

CecileRobertMichon commented:
looks like agent pools just took longer than expected 

/retest

CecileRobertMichon commented:
rebased + squashed 

@alexeldeib @jackfrancis @sonasingh46 @Jont828 PTAL

CecileRobertMichon commented:
/retest

CecileRobertMichon commented:
rebased to fix merge conflict

mboersma commented on azure/scope/managedcontrolplane.go:
```diff
@@ -431,43 +422,19 @@ func (s *ManagedControlPlaneScope) ManagedClusterSpec() (azure.ManagedClusterSpe
 
 	if clusterNetwork := s.Cluster.Spec.ClusterNetwork; clusterNetwork != nil {
 		if clusterNetwork.Services != nil {
-			// A user may provide zero or one CIDR blocks. If they provide an empty array,
```
Are we doing these "too many CIDRs" checks elsewhere now?

mboersma commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+	AdminGroupObjectIDs []string
+}
+
+// AddonProfile - Profile of managed cluster add-on.
```
```suggestion
// AddonProfile is the profile of a managed cluster add-on.
```

mboersma commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+	Enabled bool
+}
+
+// SKU - AKS SKU.
```
```suggestion
// SKU is an AKS SKU.
```

mboersma commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+	Tier string
+}
+
+// LoadBalancerProfile - Profile of the cluster load balancer.
```
```suggestion
// LoadBalancerProfile is the profile of the cluster load balancer.
```

mboersma commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+
+// SKU - AKS SKU.
+type SKU struct {
+	// Tier - Tier of a managed cluster SKU.
```
I'm pedantic enough to flag all these, but I prefer complete sentences for doc comments.

CecileRobertMichon replied:
I agree - I have to admit I just lift and shifted those as-is from types.go. Will cleanup the comments.

mboersma commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+			DNSPrefix:         &s.Name,
+			KubernetesVersion: &s.Version,
+			LinuxProfile: &containerservice.LinuxProfile{
+				AdminUsername: to.StringPtr("azureuser"),
```
Should this be a string constant?

CecileRobertMichon replied:
yes it should be

mboersma reviewed (approved):
This is a huge refactoring. I reviewed it all and I get the gist of the changes and didn't see anything obviously wrong.

shysank reviewed (commented):
looks good overall with one small clarification on AzureManagedControlPlane validation.

CecileRobertMichon commented on exp/api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -365,6 +368,74 @@ func (m *AzureManagedControlPlane) validateAPIServerAccessProfile() error {
+
+	if err := cli.Get(ctx, key, ownerCluster); err != nil {
+		if azure.ResourceNotFound(err) {
+			return nil
```
That's a good point. I had copied the pattern from https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/55a16d82fb5759e44ca1e1fa8601f778a5f23b46/exp/api/v1beta1/azuremanagedmachinepool_webhook.go#L235 so we should make both consistent if we do decide to change this. I wonder if we should retry a couple of times before failing in case the AzureManagedCluster and the Cluster are created simultaneously. Do you know if kubectl apply -f template.yaml is serial? Will it block creation if the AzureManagedControlPlane is applied before the Cluster or will it proceed to create the Cluster? 🤔

shysank replied:
> so we should make both consistent if we do decide to change this.
+1 to making it consistent. I'm okay with merging this one as is, and create a followup issue to address this concern.

> I wonder if we should retry a couple of times before failing
I'm a bit hesitant about this because we don't want to add too much complexity in validation webhook.

> Will it block creation if the AzureManagedControlPlane is applied before the Cluster or will it proceed to create the Cluster?
I think it's serial. However, it wouldn't block creation if AzureManagedControlPlane is defined before Cluster in the yaml.

CecileRobertMichon replied:
I ran some tests and it definitely works fine and doesn't block creation even if AzureManagedControlPlane is defined before Cluster

Changed both to return an error if Cluster is not found

CecileRobertMichon commented:
/retest

Jont828 commented on azure/scope/managedmachinepool.go:
```diff
@@ -0,0 +1,250 @@
+	}
+
+	replicas := int32(1)
+	if machinePool.Spec.Replicas != nil {
```
What situation would we have `machinePool.Spec.Replicas` be nil? And if so, is there any particular reason why we want to set the default replicas to 1?

CecileRobertMichon replied:
this (and the entire file) was copied directly as-is from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2168/files#diff-cfd94adbc2ff767742e457e64836e93e88c6e03c3a64c04619c54b7ed76f284eL574

Jont828 commented on azure/scope/managedmachinepool.go:
```diff
@@ -0,0 +1,250 @@
+		SKU:           managedMachinePool.Spec.SKU,
+		Replicas:      replicas,
+		Version:       normalizedVersion,
+		VnetSubnetID: azure.SubnetID(
```
Nit: could we rename `VnetSubnetID` to `SubnetID`?

Jont828 commented on azure/scope/managedmachinepool_test.go:
```diff
@@ -0,0 +1,561 @@
+				Replicas:     1,
+				Mode:         "System",
+				Cluster:      "cluster1",
+				VnetSubnetID: "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups//providers/Microsoft.Network/virtualNetworks//subnets/",
```
```suggestion
				VnetSubnetID: "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups//providers/Microsoft.Network/virtualNetworks/subnets/",
```

Remove double `//`

CecileRobertMichon replied:
I think it's empty because vnet name wasn't set, same with rg name. This was also a file split from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2168/files/5dddbd410181418a2e1f95fd83a11b6dc894c6c7#diff-a851ceb007616d5d0254ac740f53cfb4dd97e16b30629e5d201ef40527d70c82L115

Jont828 commented on azure/services/managedclusters/managedclusters.go:
```diff
@@ -18,37 +18,25 @@ package managedclusters
-	defaultUser     = "azureuser"
-	managedIdentity = "msi"
-)
+const serviceName = "managedcluster"
```
Is there a reason we're renaming this service? If so, do we want to rename the folder under `azure/services` for consistency?

CecileRobertMichon replied:
If I remember correctly, this is just so logs are a bit more coherent since we're only creating one managed cluster per AzureManagedControlPlane, so instead of printing "managedclusters creating or updating" it will print "managedcluster creating or updating" in Conditions. Not a huge deal, I didn't want to rename the package since that's a bigger change but I can revert it if you think it's not worth it. I think groups are the same if that helps.

Jont828 commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+
+var _ azure.ResourceSpecGetterWithHeaders = (*ManagedClusterSpec)(nil)
+
+// ResourceName returns the name of the group.
```
```suggestion
// ResourceName returns the name of the managed cluster.
```
or
```suggestion
// ResourceName returns the name of the AKS cluster.
```

Jont828 commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the group.
```
```suggestion
// ResourceGroupName returns the name of the resource group.
```

Jont828 commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+}
+
+// ResourceGroupName returns the name of the group.
+// Note that it is the same as the resource name in this case.
```
```suggestion

```

Jont828 commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+	return s.ResourceGroup
+}
+
+// OwnerResourceName is a no-op for groups.
```
```suggestion
// OwnerResourceName is a no-op for managed clusters.
```

Jont828 commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+	return s.Headers
+}
+
+// Parameters returns the parameters for the group.
```
```suggestion
// Parameters returns the parameters for the managed cluster.
```

Jont828 commented on azure/services/managedclusters/spec_test.go:
```diff
@@ -0,0 +1,208 @@
+				g.Expect(result.(containerservice.ManagedCluster).KubernetesVersion).To(Equal(to.StringPtr("v1.22.99")))
+			},
+		},
+		// TODO (cecile): add tests for specific parameters
```
Is this TODO done?

Jont828 commented on exp/api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -43,10 +48,9 @@ func (m *AzureManagedControlPlane) SetupWebhookWithManager(mgr ctrl.Manager) err
-
 // Default implements webhook.Defaulter so a webhook will be registered for the type.
-func (m *AzureManagedControlPlane) Default() {
+func (m *AzureManagedControlPlane) Default(_ client.Client) {
```
Why did we add an unused parameter here? Is it to fulfill an interface where it's needed somewhere else?

CecileRobertMichon replied:
correct https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/util/webhook/mutator.go#L32

we're just not using the client for default here, just for validate

Jont828 commented on azure/scope/managedcontrolplane.go:
```diff
@@ -51,12 +48,15 @@ import (
+	ManagedMachinePools []ManagedMachinePool
+}
+
+type ManagedMachinePool struct {
```
Can we move this to `managedmachinepool.go` since we're trying to separate the scopes? Or would that cause an import cycle?

Jont828 commented on exp/api/v1beta1/azuremanagedcontrolplane_webhook_test.go:
```diff
@@ -240,11 +243,11 @@ func TestValidatingWebhook(t *testing.T) {
 			g := NewWithT(t)
 			t.Parallel()
-
+			var client client.Client
```
Is this client supposed to be uninitialized?

Jont828 commented on azure/services/managedclusters/spec.go:
```diff
@@ -0,0 +1,430 @@
+	EnablePrivateClusterPublicFQDN *bool
+}
+
+var _ azure.ResourceSpecGetterWithHeaders = (*ManagedClusterSpec)(nil)
```
What's this line for?

CecileRobertMichon replied:
it's validating that the managed cluster spec implements the `ResourceSpecGetterWithHeaders` interface

it's a common go pattern to check that a type implements an interface https://golangbyexample.com/go-check-if-type-implements-interface/

Jont828 commented on azure/services/managedclusters/managedclusters.go:
```diff
@@ -184,240 +70,57 @@ func (s *Service) Reconcile(ctx context.Context) error {
+	defer cancel()
+
+	managedClusterSpec := s.Scope.ManagedClusterSpec(ctx)
+	if managedClusterSpec == nil {
```
Maybe add a log statement similar to the other async services?

CecileRobertMichon replied:
don't think we're logging anything for other services if there is no spec? It just means that service is not applicable for that object

Jont828 commented on azure/services/managedclusters/managedclusters.go:
```diff
@@ -184,240 +70,57 @@ func (s *Service) Reconcile(ctx context.Context) error {
-		return azure.WithTransientError(errors.Wrap(err, "failed to fetch existing managed cluster"), 20*time.Second)
+	managedClusterSpec := s.Scope.ManagedClusterSpec(ctx)
+	if managedClusterSpec == nil {
+		return nil
```
Maybe add a log statement like in the other async services?

CecileRobertMichon replied:
ditto

Jont828 commented on azure/services/managedclusters/managedclusters.go:
```diff
@@ -184,240 +70,57 @@ func (s *Service) Reconcile(ctx context.Context) error {
-			managedCluster.NetworkProfile.LoadBalancerProfile.OutboundIPs = &containerservice.ManagedClusterLoadBalancerProfileOutboundIPs{
-				PublicIPs: convertToResourceReferences(managedClusterSpec.LoadBalancerProfile.OutboundIPs),
+		// Update control plane endpoint.
+		if managedCluster.ManagedClusterProperties != nil && managedCluster.ManagedClusterProperties.Fqdn != nil {
```
Would it be unexpected for the properties to be nil? If so maybe we could add a log.

CecileRobertMichon replied:
not sure, I kept the same logic that was there before https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2168/files#diff-ea46b89333855bb06ff20351c648ad0a792c9fc36689b982e871b9f6fee2ef7aL366

CecileRobertMichon replied:
removed this and the pointer dereference

CecileRobertMichon commented:
@Jont828 thanks for the review! I've addressed most of your comments, PTAL. I kept the new changes as a separate commit, let me know if this looks good and I can squash.

CecileRobertMichon commented:
/retest

CecileRobertMichon commented:
@Jont828 @shysank did you have any other comments? Let me know when I can squash

CecileRobertMichon commented:
@Jont828 @shysank did you have any other comments? Let me know when I can squash

Jont828 commented on exp/api/v1beta1/azuremanagedcontrolplane_webhook.go:
```diff
@@ -270,7 +270,7 @@ func (m *AzureManagedControlPlane) Validate() error {
 }
 
 // validate DNSServiceIP.
```
Nit: capitalize "validate" in comment to be consistent

CecileRobertMichon replied:
I fixed the comment to be consistent with the function name, and fixed the others that were using inconsistent capitalization

Jont828 reviewed (commented):
Small thing but other than that I think we're good to go!

CecileRobertMichon commented:
Squashed

CecileRobertMichon commented:
@Jont828 @mboersma @shysank PTAL, addressed all comments and squashed

Jont828 commented:
/lgtm

CecileRobertMichon commented:
@mboersma @shysank any objections on merging this one?

mboersma commented:
lgtm! Please merge, your patience is admirable.

mboersma commented:
/approve
